File: src/omnia/data/structure/observable/WritableObservableListImpl.java
package omnia.data.structure.observable;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;
import java.util.Iterator;
import java.util.OptionalInt;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;
import omnia.data.iterate.IntegerRangeIterator;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;

final class WritableObservableListImpl<E> implements WritableObservableList<E> {

  private volatile ImmutableList<E> currentState = ImmutableList.empty();
  private final Subject<MutationEvent> mutationEventSubject = PublishSubject.create();
  private final ObservableChannels observableChannels = new ObservableChannels();

  @Override
  public void insertAt(int index, E item) {
    IndexRange range = IndexRange.just(index);
    mutateState(
        state -> true,
        state -> state.toBuilder().insertAt(index, item).build(),
        (previousState, currentState) -> generateInsertAtMutations(currentState, range));
  }

  @Override
  public void removeAt(int index) {
    IndexRange range = IndexRange.just(index);
    mutateState(
        state -> true,
        state -> state.toBuilder().removeAt(index).build(),
        (previousState, currentState) -> generateRemoveAtMutations(previousState, range));
  }

  @Override
  public void replaceAt(int index, E item) {
    IndexRange range = IndexRange.just(index);
    mutateState(
        state -> true,
        state -> state.toBuilder().replaceAt(index, item).build(),
        (previousState, currentState) ->
            generateReplaceAtMutations(previousState, currentState, range));
  }

  @Override
  public void add(E item) {
    mutateState(
        state -> true,
        state -> state.toBuilder().add(item).build(),
        (previousState, currentState) ->
            generateInsertAtMutations(currentState, IndexRange.just(currentState.count() - 1)));
  }

  @Override
  public boolean remove(Object item) {
    synchronized(this) {
      OptionalInt index = getState().indexOf(item);
      return mutateState(
          state -> index.isPresent(),
          state -> state.toBuilder().removeAt(index.getAsInt()).build(),
          (previousState, currentState) ->
              generateRemoveAtMutations(
                  previousState, IndexRange.just(index.getAsInt())));
    }
  }

  @Override
  public void clear() {
    synchronized(this) {
      mutateState(
          ImmutableList::isPopulated,
          state -> ImmutableList.empty(),
          (previousState, currentState) ->
              generateRemoveAtMutations(
                  previousState, new IndexRange(0, previousState.count())));
    }
  }

  private boolean mutateState(
      Predicate<ImmutableList<E>> shouldMutate,
      Function<ImmutableList<E>, ImmutableList<E>> mutator,
      BiFunction<ImmutableList<E>, ImmutableList<E>, List<ListOperation<E>>> mutationsGenerator) {
    synchronized(this) {
      ImmutableList<E> previousState = currentState;
      if (!shouldMutate.test(previousState)) {
        return false;
      }
      ImmutableList<E> newState = mutator.apply(previousState);
      currentState = newState;
      mutationEventSubject.onNext(new MutationEvent(newState, mutationsGenerator.apply(previousState, newState)));
      return true;
    }
  }

  private static <E> ImmutableList<ListOperation<E>> generateInsertAtMutations(
      ImmutableList<E> state, IndexRange range) {
    ImmutableList.Builder<ListOperation<E>> builder = ImmutableList.builder();

    // Move must be done before insert to make way for new items
    if (range.end() < state.count()) {
      int numItemsMoved = state.count() - range.end();
      IndexRange moveFromRange = new IndexRange(range.start(), range.start() + numItemsMoved);
      IndexRange moveToRange = new IndexRange(range.end(), range.end() + numItemsMoved);
      builder.add(new MoveInList<>(moveToRange.asSublistOf(state), moveFromRange, moveToRange));
    }

    builder.add(new AddToList<>(range.asSublistOf(state), range));

    return builder.build();
  }

  private static <E> ImmutableList<ListOperation<E>> generateRemoveAtMutations(
      ImmutableList<E> previousState, IndexRange range) {
    ImmutableList.Builder<ListOperation<E>> builder = ImmutableList.builder();

    /// Removal must be done BEFORE move to make space for moved items
    builder.add(new RemoveFromList<>(range.asSublistOf(previousState), range));

    if (range.end() < previousState.count()) {
      int numItemsMoved = previousState.count() - range.end();
      IndexRange moveFromRange = new IndexRange(range.end(), range.end() + numItemsMoved);
      IndexRange moveToRange = new IndexRange(range.start(), range.start() + numItemsMoved);
      builder.add(
          new MoveInList<>(moveFromRange.asSublistOf(previousState), moveFromRange, moveToRange));
    }

    return builder.build();
  }

  private static <E> ImmutableList<ListOperation<E>> generateReplaceAtMutations(
      ImmutableList<E> previousState, ImmutableList<E> currentState, IndexRange range) {
    return ImmutableList.of(
        new ReplaceInList<>(
            range.asSublistOf(previousState), range.asSublistOf(currentState), range));
  }

  @Override
  public Iterator<E> iterator() {
    return getState().iterator();
  }

  @Override
  public boolean contains(Object element) {
    return getState().contains(element);
  }

  @Override
  public int count() {
    return getState().count();
  }

  @Override
  public boolean isPopulated() {
    return getState().isPopulated();
  }

  @Override
  public E itemAt(int index) {
    return getState().itemAt(index);
  }

  @Override
  public OptionalInt indexOf(Object item) {
    return getState().indexOf(item);
  }

  @Override
  public Stream<E> stream() {
    return getState().stream();
  }

  @Override
  public ObservableList<E> toReadOnly() {
    return new ObservableList<>() {
      @Override
      public ObservableChannels<E> observe() {
        return WritableObservableListImpl.this.observe();
      }

      @Override
      public Iterator<E> iterator() {
        return WritableObservableListImpl.this.iterator();
      }

      @Override
      public boolean contains(Object element) {
        return WritableObservableListImpl.this.contains(element);
      }

      @Override
      public int count() {
        return WritableObservableListImpl.this.count();
      }

      @Override
      public E itemAt(int index) {
        return WritableObservableListImpl.this.itemAt(index);
      }

      @Override
      public OptionalInt indexOf(Object item) {
        return WritableObservableListImpl.this.indexOf(item);
      }

      @Override
      public Stream<E> stream() {
        return WritableObservableListImpl.this.stream();
      }
    };
  }

  @Override
  public ObservableChannels observe() {
    return observableChannels;
  }

  private ImmutableList<E> getState() {
    synchronized(this) {
      return currentState;
    }
  }

  private static final class AddToList<E> implements ObservableList.AddToList<E> {
    private final List<E> items;
    private final IndexRange indices;

    private AddToList(List<E> items, IndexRange indices) {
      this.items = items;
      this.indices = indices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IndexRange indices() {
      return indices;
    }
  }

  private static final class MoveInList<E> implements ObservableList.MoveInList<E> {
    private final List<E> items;
    private final IndexRange previousIndices;
    private final IndexRange currentIndices;

    private MoveInList(List<E> items, IndexRange previousIndices, IndexRange currentIndices) {
      this.items = items;
      this.previousIndices = previousIndices;
      this.currentIndices = currentIndices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IndexRange previousIndicies() {
      return previousIndices;
    }

    @Override
    public IndexRange currentIndicies() {
      return currentIndices;
    }
  }

  private static final class RemoveFromList<E> implements ObservableList.RemoveFromList<E> {
    private final List<E> items;
    private final IndexRange indices;

    private RemoveFromList(List<E> items, IndexRange indices) {
      this.items = items;
      this.indices = indices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IndexRange indices() {
      return indices;
    }
  }

  private static final class ReplaceInList<E> implements ObservableList.ReplaceInList<E> {
    private final List<E> replacedItems;
    private final List<E> newItems;
    private final IndexRange indices;

    private ReplaceInList(List<E> replacedItems, List<E> newItems, IndexRange indices) {
      this.replacedItems = replacedItems;
      this.newItems = newItems;
      this.indices = indices;
    }

    @Override
    public List<E> replacedItems() {
      return replacedItems;
    }

    @Override
    public List<E> newItems() {
      return newItems;
    }

    @Override
    public IndexRange indices() {
      return indices;
    }
  }

  private static final class IndexRange implements ObservableList.IndexRange {
    private final int start;
    private final int end;

    private static IndexRange just(int index) {
      return new IndexRange(index, index + 1);
    }

    private IndexRange(int start, int end) {
      if (start <= end) {
        throw new IllegalArgumentException(
            "range must describe a non-empty set of integers. start=" + start + " end=" + end);
      }
      this.start = start;
      this.end = end;
    }

    @Override
    public int start() {
      return start;
    }

    @Override
    public int end() {
      return end;
    }

    @Override
    public int endInclusive() {
      return end - 1;
    }

    @Override
    public int count() {
      return end - start;
    }

    @Override
    public boolean isPopulated() {
      return start < end;
    }

    @Override
    public Iterator<Integer> iterator() {
      return IntegerRangeIterator.create(start, end);
    }

    private <E> ImmutableList<E> asSublistOf(ImmutableList<E> list) {
      return list.sublistStartingAt(start()).to(end());
    }
  }

  private class ObservableChannels extends GenericObservableChannels<List<E>, MutationEvent>
      implements ObservableList.ObservableChannels<E> {


    protected ObservableChannels() {
      super(
          Flowable.<List<E>>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(getState());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(
                  mutationEventSubject.toFlowable(BackpressureStrategy.BUFFER)
                      .map(MutationEvent::state)),
          Flowable.<MutationEvent>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(generateMutationEventForNewSubscription());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(mutationEventSubject.toFlowable(BackpressureStrategy.BUFFER)));
    }
  }

  private class MutationEvent
      extends GenericMutationEvent<List<E>, List<ObservableList.ListOperation<E>>>
      implements ObservableList.MutationEvent<E> {
    private MutationEvent(List<E> state, List<ListOperation<E>> operations) {
      super(state, operations);
    }
  }

  private MutationEvent generateMutationEventForNewSubscription() {
    List<E> state = getState();
    return new MutationEvent(
        state, ImmutableList.of(new AddToList<>(state, new IndexRange(0, state.count()))));
  }
}

File: omnia/src/commonMain/omnia/data/structure/observable/writable/WritableObservableListImpl.kt
package omnia.data.structure.observable.writable

import com.badoo.reaktive.observable.concatWith
import com.badoo.reaktive.observable.map
import com.badoo.reaktive.observable.observable
import com.badoo.reaktive.subject.publish.PublishSubject
import kotlin.concurrent.Volatile
import omnia.data.structure.IntRange
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import omnia.data.structure.observable.ObservableList
import omnia.data.structure.observable.ObservableList.ListOperation

internal class WritableObservableListImpl<E : Any> : WritableObservableList<E> {

  @Volatile
  private var currentState = ImmutableList.empty<E>()
  private val mutationEvents = PublishSubject<MutationEvent>()

  override fun insertAt(index: Int, item: E) {
    val range: IntRange = IntRange.just(index)
    mutateState(
      { true },
      { state -> state.toBuilder().insertAt(index, item).build() }
    ) { _, currentState -> generateInsertAtMutations(currentState, range) }
  }

  override fun removeAt(index: Int): E {
    val range: IntRange = IntRange.just(index)
    val value: E = state.itemAt(index)
    mutateState(
      { true },
      { state -> state.toBuilder().removeAt(index).build() }
    ) { previousState: ImmutableList<E>, _: ImmutableList<E> ->
      generateRemoveAtMutations(
        previousState,
        range
      )
    }
    return value
  }

  override fun replaceAt(index: Int, item: E): E {
    val range: IntRange = IntRange.just(index)
    val value: E = state.itemAt(index)
    mutateState(
      { true },
      { state -> state.toBuilder().replaceAt(index, item).build() }
    ) { previousState, currentState ->
      generateReplaceAtMutations(
        previousState,
        currentState,
        range
      )
    }
    return value
  }

  override fun add(item: E) {
    mutateState(
      { true },
      { state -> state.toBuilder().add(item).build() }
    ) { _, currentState ->
      generateInsertAtMutations(
        currentState,
        IntRange.just(currentState.count - 1)
      )
    }
  }

  override fun addAll(items: Iterable<E>) {
    mutateState(
      { true },
      { state -> state.toBuilder().addAll(items).build() }
    ) { previousState, currentState ->
      generateInsertAtMutations(
        currentState,
        IntRange.startingAt(previousState.count).endingAt(currentState.count)
      )
    }
  }

  override fun removeUnknownTyped(item: Any?): Boolean {
    val index = state.indexOf(item)
    return mutateState(
      { index != null },
      { state -> state.toBuilder().removeAt(index!!).build() }
    ) { previousState, _ ->
      generateRemoveAtMutations(
        previousState, IntRange.just(index!!)
      )
    }
  }

  override fun clear() {
    mutateState({ obj: ImmutableList<E> -> obj.isPopulated },
      { ImmutableList.empty() }
    ) { previousState, _ ->
      generateRemoveAtMutations(
        previousState, IntRange.startingAt(0).endingAt(previousState.count)
      )
    }
  }

  private fun mutateState(
    shouldMutate: (ImmutableList<E>) -> Boolean,
    mutator: (ImmutableList<E>) -> ImmutableList<E>,
    mutationsGenerator: (ImmutableList<E>, ImmutableList<E>) -> List<ListOperation<E>>
  ): Boolean {
    val previousState = currentState
    if (!shouldMutate(previousState)) {
      return false
    }
    val newState = mutator(previousState)
    currentState = newState
    mutationEvents.onNext(
      MutationEvent(newState, mutationsGenerator(previousState, newState))
    )
    return true
  }

  override fun iterator(): Iterator<E> {
    return state.iterator()
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return state.containsUnknownTyped(item)
  }

  override val count: Int
    get() {
      return state.count
    }

  override val isPopulated: Boolean
    get() = state.isPopulated

  override fun itemAt(index: Int): E {
    return state.itemAt(index)
  }

  override fun indexOf(item: Any?): Int? {
    return state.indexOf(item)
  }

  override fun toReadOnly(): ObservableList<E> {
    return object : ObservableList<E> {
      override val observables: ObservableList.Observables<E>
        get() {
          return this@WritableObservableListImpl.observables
        }

      override fun iterator(): Iterator<E> {
        return this@WritableObservableListImpl.iterator()
      }

      override fun containsUnknownTyped(item: Any?): Boolean {
        return this@WritableObservableListImpl.containsUnknownTyped(item)
      }

      override val count: Int
        get() {
          return this@WritableObservableListImpl.count
        }

      override fun itemAt(index: Int): E {
        return this@WritableObservableListImpl.itemAt(index)
      }

      override fun indexOf(item: Any?): Int? {
        return this@WritableObservableListImpl.indexOf(item)
      }
    }
  }

  override val observables = Observables()

  private val state: ImmutableList<E> get() = currentState

  private class AddToList<E : Any>(override val items: List<E>, override val indices: IntRange) :
    ObservableList.AddToList<E>

  private class MoveInList<E : Any>(
    override val items: List<E>,
    override val previousIndices: IntRange,
    override val currentIndices: IntRange
  ) : ObservableList.MoveInList<E>

  private class RemoveFromList<E : Any>(
    override val items: List<E>,
    override val indices: IntRange
  ) : ObservableList.RemoveFromList<E>

  private class ReplaceInList<E : Any>(
    override val replacedItems: List<E>,
    override val newItems: List<E>,
    override val indices: IntRange
  ) : ObservableList.ReplaceInList<E>

  inner class Observables : GenericObservables<List<E>, MutationEvent>(
    observable<List<E>> {
      it.onNext(state)
      it.onComplete()
    }
      .concatWith(mutationEvents.map { it.state }),
    observable<MutationEvent> {
      it.onNext(generateMutationEventForNewSubscription())
      it.onComplete()
    }
      .concatWith(mutationEvents)),
    ObservableList.Observables<E>

  inner class MutationEvent(state: List<E>, operations: List<ListOperation<E>>) :
    GenericMutationEvent<List<E>, List<ListOperation<E>>>(state, operations),
    ObservableList.MutationEvent<E>

  private fun generateMutationEventForNewSubscription(): MutationEvent {
    val state: List<E> = state
    return MutationEvent(
      state, ImmutableList.of(AddToList(state, IntRange.startingAt(0).endingAt(state.count)))
    )
  }

  companion object {

    private fun <E : Any> generateInsertAtMutations(
      state: ImmutableList<E>, range: IntRange
    ): ImmutableList<ListOperation<E>> {
      val builder: ImmutableList.Builder<ListOperation<E>> = ImmutableList.builder()

      // Move must be done before insert to make way for new items
      if (range.end < state.count) {
        val numItemsMoved = state.count - range.end
        val moveFromRange: IntRange = IntRange.startingAt(range.start).withLength(numItemsMoved)
        val moveToRange: IntRange = IntRange.startingAt(range.end).withLength(numItemsMoved)
        builder.add(MoveInList(state.getSublist(moveToRange), moveFromRange, moveToRange))
      }
      builder.add(AddToList(state.getSublist(range), range))
      return builder.build()
    }

    private fun <E : Any> generateRemoveAtMutations(
      previousState: ImmutableList<E>, range: IntRange
    ): ImmutableList<ListOperation<E>> {
      val builder: ImmutableList.Builder<ListOperation<E>> = ImmutableList.builder()

      /// Removal must be done BEFORE move to make space for moved items
      builder.add(RemoveFromList(previousState.getSublist(range), range))
      if (range.end < previousState.count) {
        val numItemsMoved = previousState.count - range.end
        val moveFromRange: IntRange = IntRange.startingAt(range.end).withLength(numItemsMoved)
        val moveToRange: IntRange = IntRange.startingAt(range.start).withLength(numItemsMoved)
        builder.add(
          MoveInList(previousState.getSublist(moveFromRange), moveFromRange, moveToRange)
        )
      }
      return builder.build()
    }

    private fun <E : Any> generateReplaceAtMutations(
      previousState: ImmutableList<E>, currentState: ImmutableList<E>, range: IntRange
    ): ImmutableList<ListOperation<E>> {
      return ImmutableList.of(
        ReplaceInList(
          previousState.getSublist(range), currentState.getSublist(range), range
        )
      )
    }
  }
}
